lightning/src/chain/keysinterface.rs

```diff
@@ -423,12 +433,36 @@ pub trait KeysInterface {
 	/// this trait to parse the invoice and make sure they're signing what they expect, rather than
 	/// blindly signing the hash.
 	/// The hrp is ascii bytes, while the invoice data is base32.
-	fn sign_invoice(&self, hrp_bytes: &[u8], invoice_data: &[u5]) -> Result<RecoverableSignature, ()>;
+	///
+	/// If `invoice_type` is [`Invoice::Bolt11`], then this invoice must be signed by the node secret
+	/// provided by [`KeysInterface::get_node_secret`]. Else if it is [`Invoice::Phantom`], then this
+	/// invoice must be signed by the phantom node secret provided by
+	/// [`KeysInterface::get_phantom_secret`].
+	fn sign_invoice(&self, hrp_bytes: &[u8], invoice_data: &[u5], invoice_type: Invoice) -> Result<RecoverableSignature, ()>;
 
 	/// Get secret key material as bytes for use in encrypting and decrypting inbound payment data.
 	///
+	/// If the implementor of this trait supports [phantom node payments], then every node that is
+	/// intended to be included in the phantom invoice(s) route hints must return the same value from
+	/// this method.
+	///
 	/// This method must return the same value each time it is called.
+	///
+	/// [phantom node payments]: crate::ln::channelmanager::ChannelManager::get_phantom_scid
 	fn get_inbound_payment_key_material(&self) -> KeyMaterial;
+
+	/// Get the phantom node secret key (aka node_id or network_key) for use in receiving a [phantom
+	/// node payment]. This secret key is used to sign phantom invoices.
+	///
+	/// Note that [`KeysManager`] does NOT support phantom node payments, [`PhantomKeysManager`] must
+	/// be used instead.
+	///
+	/// If the implementor of this trait supports [phantom node payments], then every node that is
+	/// intended to be included in the phantom invoice(s) route hints must return the same value from
+	/// this method.
+	///
+	/// [phantom node payment]: crate::ln::channelmanager::ChannelManager::get_phantom_scid
+	fn get_phantom_secret(&self) -> Option<SecretKey>;
```

Comment:
I definitely could be missing something here, but is there a reason why this needs to be different from `get_inbound_payment_key_material`? If they were the same, then the phantom-ness wouldn't need to bleed into the interface.

Reply:
My understanding is that reusing keys is an iffy security practice, so we're following that general principle

Reply:
Could the phantom secret be part of the `ExpandedKey` given that latter is also expanded from `get_inbound_payment_key_material`? That would achieve the same result, I think.

Reply:
No, we want to get rid of this function while having the phantom node secret key be handled by a potentially-offline signer. That way the keys required to generate an invoice are not available on the "untrusted" machine.